Updating federation up script to create secrets with federation-apiserver and k8s apiservers kubeconfigs

nikhiljindal opened the pull request:
Follow up to https://github.com/kubernetes/kubernetes/pull/26819

cc @kubernetes/sig-cluster-federation

nikhiljindal committed:
Updating federation up script to create a secret with federation-apiserver kubeconfig

nikhiljindal commented:
Am still testing it out

quinton-hoole commented:
@nikhiljindal Let me know as soon as it's ready for review.

quinton-hoole commented on federation/cluster/common.sh:
```diff
@@ -146,6 +147,21 @@ function create-federated-api-objects {
     done
     
+    # Create a kubeconfig with credentails for federation-apiserver and create a
+    # secret for it.
```
This comment does not seem to apply to the next line of code?

quinton-hoole commented:
Minor nit, otherwise LGTM.

nikhiljindal commented:
Sorry I havent been able to get this to work yet.

nikhiljindal commented:
Here is the problem I am having:
When I call create-kubeconfig with a different value to create a kubeconfig file with just the federation-apiserver credentials, create-kubeconfig exports KUBECONFIG with this new value and hence all future kubectl commands are reading this new kubeconfig file which can only talk to federation-apiserver and hence many kubectl commands are failing.
I am fixing it so that create-kubeconfig does not export KUBECONFIG with a new value.

nikhiljindal commented:
Updated the code.

I am still having trouble with pushing my images, so federation-apiserver and federation-controller-manager are not coming up for me.
But verified that the script creates the right secret and it is available for the controller manager to use.

@quinton-hoole PTAL. Should be ready to be merged

nikhiljindal commented on cluster/common.sh:
```diff
@@ -128,8 +128,16 @@ function create-kubeconfig() {
+# To explicitly name the context being removed, use OVERRIDE_CONTEXT
 function clear-kubeconfig() {
   export KUBECONFIG=${KUBECONFIG:-$DEFAULT_KUBECONFIG}
+  OVERRIDE_CONTEXT=${OVERRIDE_CONTEXT:-}
```
This is an independent change. Without this change, running `hack/e2e.go --up` followed by `hack/e2e.go --down` would leave contexts in kubeconfig that can affect later runs. This change ensures that kubeconfig is cleared correctly.

nikhiljindal commented:
@k8s-bot e2e test this issue: https://github.com/kubernetes/kubernetes/issues/23545

@k8s-bot node e2e test this issue: https://github.com/kubernetes/kubernetes/issues/26431

nikhiljindal commented:
Same errors happening again.

nikhiljindal commented:
Trying again
@k8s-bot e2e test this issue: #23545

@k8s-bot node e2e test this issue: #26431

nikhiljindal commented:
@k8s-bot node e2e test this issue: #26431

quinton-hoole commented:
LGTM!

nikhiljindal committed:
Adding secrets for kubernetes clusters as well

nikhiljindal commented:
Added another commit to create secrets for the kubernetes apiservers as well.
Also updated the logic to delete these generated secrets in cleanup.

@quinton-hoole PTAL.

nikhiljindal commented:
Note that I am creating secrets for k8s clusters only on GCE for now. 
Doing the same on vagrant seems straightforward, but I am not so sure about GKE.

In GKE script, we directly call gcloud create cluster to create the cluster: https://github.com/kubernetes/kubernetes/blob/master/cluster/gke/util.sh#L183. I guess that updates the kubeconfig as well.

nikhiljindal commented:
This PR with GCE only secrets should atleast unblock @madhusudancs for his service controller tests.

quinton-hoole commented:
Awesome.  LGTM.

quinton-hoole commented:
@k8s-oncall @yujuhong could we merge this manually please?  It's passed all checks, but the [http://submit-queue.k8s.io/#/queue](submit-queue) is currently stalled for unrelated reasons.  This PR is blocking further e2e testing of cluster federation, which is in turn blocking the v1.3 launch.  Thanks!

quinton-hoole commented:
Subsequent unit test failure looks like a flake. https://github.com/kubernetes/kubernetes/issues/27024

@k8s-bot test this issue: #27024

madhusudancs commented:
@nikhiljindal Thank you very much!

quinton-hoole commented:
Manually merging.  See note above for rationale,
